atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/AbortedTransactionSoftCache.java

```diff
@@ -0,0 +1,191 @@
+                .collectToMap();
+    }
+
+    private long getLatestTsSeenSoFar(Set<Long> batch, Optional<PatchyCache> maybeSnapshot) {
+        long latestTsSeenSoFar = batch.stream().max(Comparator.naturalOrder()).orElse(0L);
+
+        if (maybeSnapshot.isPresent()) {
+            latestTsSeenSoFar = Math.max(latestTsSeenSoFar, maybeSnapshot.get().lastKnownConcludedTimestamp);
+        }
+
+        long latestBucket = AbortedTimestampUtils.getBucket(latestTsSeenSoFar);
+        // The purpose of this call is to refresh the knownConcluded store for current bucket if it is not up-to-date.
+        // Do not remove this line without considering perf implications.
+        knownConcludedTransactions.isKnownConcluded(
+                AbortedTimestampUtils.getMaxTsInCurrentBucket(latestBucket),
+                KnownConcludedTransactions.Consistency.REMOTE_READ);
+
+        latestTsSeenSoFar =
+                Math.max(latestTsSeenSoFar, knownConcludedTransactions.lastLocallyKnownConcludedTimestamp());
+        return latestTsSeenSoFar;
+    }
+
+    private PatchyCache refreshPatchyCache(
+            Optional<PatchyCache> maybeSnapshot, long latestTsSeenSoFar, long latestBucketSeenSoFar) {
+        PatchyCache refreshedPatchyCache;
+
+        if (maybeSnapshot.isEmpty() || maybeSnapshot.get().bucket < latestBucketSeenSoFar) {
+            refreshedPatchyCache = loadPatchyBucket(latestTsSeenSoFar, latestBucketSeenSoFar);
+        } else {
+            // need to load a range of timestamps
+            refreshedPatchyCache = extendPatchIfNeeded(maybeSnapshot.get(), latestTsSeenSoFar);
+        }
+
+        patchyCache = refreshedPatchyCache;
+        return refreshedPatchyCache;
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    private PatchyCache extendPatchIfNeeded(PatchyCache snapshot, long latestTs) {
+        long currentLastKnownConcluded = snapshot.lastKnownConcludedTimestamp;
+
+        // It is possible that cache was updated by the time we reach here and refresh is not required any more.
+        if (latestTs <= currentLastKnownConcluded) {
+            return snapshot;
+        }
+
+        Set<Long> newAbortedTransactions =
+                futileTimestampStore.getAbortedTransactionsInRange(currentLastKnownConcluded, latestTs);
+        snapshot.extend(latestTs, newAbortedTransactions);
+        return snapshot;
+    }
+
+    private PatchyCache loadPatchyBucket(long latestTsSeenSoFar, long latestBucketSeenSoFar) {
+        long maxTsInCurrentBucket = AbortedTimestampUtils.getMaxTsInCurrentBucket(latestBucketSeenSoFar);
+
+        Set<Long> futileTimestamps = futileTimestampStore.getAbortedTransactionsInRange(
+                AbortedTimestampUtils.getMinTsInBucket(latestBucketSeenSoFar), maxTsInCurrentBucket);
+
+        return new PatchyCache(Math.min(latestTsSeenSoFar, maxTsInCurrentBucket), futileTimestamps);
+    }
+
+    private Optional<PatchyCache> getSnapshot() {
```

Comment:
nit: leave a note somewhere that this is a mutable snapshot, but _guaranteed to be for the correct bucket_ (snapshot gives the idea that it's immutable)